corehq/celery_monitoring/heartbeat.py

```diff
@@ -0,0 +1,67 @@
+from __future__ import absolute_import
+from __future__ import print_function
+
+import datetime
+
+from celery.task import periodic_task
+
+from corehq.util.datadog.gauges import datadog_gauge
+from corehq.util.quickcache import quickcache
+
+
+HEARTBEAT_FREQUENCY = datetime.timedelta(seconds=10)
+HEARTBEAT_CACHE_TIMEOUT = datetime.timedelta(days=2)
+
+
+class HeartbeatNeverRecorded(Exception):
+    pass
+
+
+class Heartbeat(object):
+    def __init__(self, queue):
+        self.queue = queue
+
+    @quickcache(['self.queue'], timeout=HEARTBEAT_CACHE_TIMEOUT.total_seconds())
+    def get_last_seen(self):
+        # This function relies on the cache getting set manually in mark_seen
+        raise HeartbeatNeverRecorded()
+
+    def mark_seen(self):
+        self.get_last_seen.set_cached_value(self).to(datetime.datetime.utcnow())
+
+    def get_blockage_duration(self):
+        """
+        Estimate the amount of time the queue has been blocked
+
+        :returns a timedelta or raises HeartbeatNeverRecorded
+        """
+        # Subtract off the time between heartbeats
+        # since we don't know how long it's been since the last heartbeat
+        return max(datetime.datetime.utcnow() - self.get_last_seen() - HEARTBEAT_FREQUENCY,
+                   datetime.timedelta(seconds=0))
+
+    @property
+    def periodic_task_name(self):
+        return 'heartbeat__{}'.format(self.queue)
+
+    @property
+    def periodic_task(self):
+        queue = self.queue
+
+        def heartbeat():
+            hb = Heartbeat(queue)
```

Comment:
Is it necessary to create a new instance here? Could it just use `self` instead?

Reply:
Yeah it may not be necessary. I wanted to avoid having this task function be a closure (or rather rely on the context it gets from being a closure), just because that felt weird to me. I think given how this gets instantiated on the celery worker etc that it would actually work just fine like you're saying, but it felt a little harder to reason with to me for some reason so I did i this way

Reply:
> I wanted to avoid having this task function be a closure (or rather rely on the context it gets from being a closure)...

Isn't it doing exactly that by using `queue` on this line?

Aside: the thing that makes this class hard to reason about for me is the non-standard use of "caching" (which doesn't really feel like caching in this context), as @emord pointed out.

Reply:
Yeah you're right I think while in my mind that was somehow less entangled than reusing self, it maybe isn't given the way closures work. I can change this to use `self`. I will probably rename it ouside the closure (`hb = self`) just because `self` has a different conventional meaning inside a celery task, but see what you're saying that there's no need to create a new `Heartbeat` instance. I'll implement this.

Reply:
Eh, screw it I'll just use self. I think what I was suggesting causes more confusion than it avoids. I can worry about it if we ever need to introduce a celery `self` arg.